### PR TITLE
Poc grouping expenses

### PR DESCRIPTION
**Pull request by ruuushhh**

### CODE REVIEW

File: apps/fyle/models.py [hunk 1]
```diff
@@ -359,16 +360,53 @@ class Meta:
...
+                        grouped_data[spent_at].append(expense)
+                    grouped_expenses = list(grouped_data.values())
+                    filtered_expenses = []
+                    for expense_group in grouped_expenses:
+                        total_amount=0
+                        for expense in expense_group:
+                            total_amount += expense.amount
+                        if total_amount < 0:
+                            expense_group = list(filter(lambda expense: expense.amount > 0, expense_group))
+                        filtered_expenses.extend(expense_group)
```
Comment (ashwin1111) [inline apps/fyle/models.py:395]: we're adding expense group to this array, but all other places we're doing expenses

File: apps/fyle/models.py [hunk 2]
```diff
@@ -359,16 +360,53 @@ class Meta:
...
+        expense_groups = []
         expense_group_settings = ExpenseGroupSettings.objects.get(workspace_id=workspace_id)
 
         reimbursable_expense_group_fields = expense_group_settings.reimbursable_expense_group_fields
         reimbursable_expenses = list(filter(lambda expense: expense.fund_source == 'PERSONAL', expense_objects))
 
-        expense_groups = _group_expenses(reimbursable_expenses, reimbursable_expense_group_fields, workspace_id)
+        reimbursable_expense_groups = _group_expenses(reimbursable_expenses, reimbursable_expense_group_fields, workspace_id)
+
+        for reimbursable_expense_group in reimbursable_expense_groups:
```
[the remark this answers is not in the document]
Comment (ruuushhh) [reply apps/fyle/models.py:375]: +1

File: apps/fyle/models.py [hunk 3]
```diff
@@ -338,6 +339,37 @@ def _group_expenses(expenses, group_fields, workspace_id):
     return expense_groups
 
 
+def filter_expenses(expense_groups, expenses: Expense, expenses_object, expense_group_fields):
```
Comment (ashwin1111) [inline apps/fyle/models.py:342]: filter_expense_groups()

File: apps/fyle/models.py [hunk 4]
```diff
@@ -338,6 +339,37 @@ def _group_expenses(expenses, group_fields, workspace_id):
...
+            expenses_object == 'EXPENSE_REPORT'
+            and 'expense_id' not in expense_group_fields
+        ):
+            total_amount = sum(expense.amount for expense in filtered_expenses)
+
+            if total_amount < 0:
+                filtered_expenses = list(filter(lambda expense: expense.amount > 0, filtered_expenses))
+        
+        # Export type => Journal Entry, Expense Report and Group By => Expense
+        elif (expenses_object in ('JOURNAL_ENTRY', 'EXPENSE_REPORT') and 'expense_id' in expense_group_fields) or expenses_object == 'JOURNAL_ENTRY':
```
Comment (ashwin1111) [inline apps/fyle/models.py:361]: `elif (expenses_object == 'EXPENSE_REPORT' and 'expense_id' in expense_group_fields) or expenses_object == 'JOURNAL_ENTRY':`

File: apps/fyle/models.py [hunk 5]
```diff
@@ -338,6 +339,37 @@ def _group_expenses(expenses, group_fields, workspace_id):
...
+            filtered_expenses = list(filter(lambda expense: expense.amount > 0, filtered_expenses))
```
Comment (ashwin1111) [inline apps/fyle/models.py:362]: this is used twice, move it to a separate func and call it in both places

Comment (ashwin1111) [reply apps/fyle/models.py:362]: filter_negative_expenses()

**Review by ashwin1111 — APPROVED** [review]

File: apps/fyle/models.py [hunk 6]
```diff
@@ -338,6 +339,38 @@ def _group_expenses(expenses, group_fields, workspace_id):
     return expense_groups
 
 
+def filter_nagative_expenses(filtered_expenses):
```
Comment (Shwetabhk) [inline apps/fyle/models.py:342]:
```suggestion
def filter_negative_expenses(filtered_expenses):
```

File: apps/fyle/models.py [hunk 7]
```diff
@@ -338,6 +339,38 @@ def _group_expenses(expenses, group_fields, workspace_id):
...
+        # Export type => Expense Report and Group By => Report
+        if expenses_object == 'EXPENSE_REPORT' and 'expense_id' not in expense_group_fields:
+            total_amount = sum(expense.amount for expense in filtered_expenses)
+
+            if total_amount < 0:
+                filtered_expenses = filter_nagative_expenses(filtered_expenses)
+        
+        # Export type => Journal Entry, Expense Report and Group By => Expense
+        elif (expenses_object == 'EXPENSE_REPORT' and 'expense_id' in expense_group_fields) or expenses_object == 'JOURNAL_ENTRY':
+            filtered_expenses = filter_nagative_expenses(filtered_expenses)
```
Comment (Shwetabhk) [inline apps/fyle/models.py:363]:
```suggestion
            filtered_expenses = filter_negative_expenses(filtered_expenses)
```

**Review by Shwetabhk — APPROVED** [review]

**Pull request merged**